From 37b9510c08ba4d2c01f2fa83690a4db2419bd951 Mon Sep 17 00:00:00 2001
From: Jason Chen <b02280@freescale.com>
Date: Tue, 14 Sep 2010 13:37:28 +0800
Subject: [PATCH 127/249] ENGR00131417 ipu lib: change FG position setting
 code

There is one bug related with FG position setting.
Detailed reproduce steps:
1. bootup system on WVGA lcd.
2. play video on FG on WVGA based on ipu lib.
3. DP-DC switch by below cmd (enable both WVGA and DVI 720P):
echo 4 > /sys/class/graphics/fb2/blank
echo 4 > /sys/class/graphics/fb0/blank
echo 4 > /sys/class/graphics/fb1/blank
echo 1-layer-fb > /sys/class/graphics/fb0/fsl_disp_property
echo U:1280x720p-60 > /sys/class/graphics/fb1/mode
echo 0 > /sys/class/graphics/fb1/blank
echo 0 > /sys/class/graphics/fb0/blank
4. stop video play by ctrl-C
5. play video again on DVI
6. system will hang.

Move set position to 0,0 into driver and remove it in ipu lib.
Set first position after fg unblank.

Signed-off-by: Jason Chen <b02280@freescale.com>
---
 ipu/mxc_ipu_hl_lib.c | 18 +++++++-----------
 1 file changed, 7 insertions(+), 11 deletions(-)

diff --git a/ipu/mxc_ipu_hl_lib.c b/ipu/mxc_ipu_hl_lib.c
index 58b8f63..0911fbd 100644
--- a/ipu/mxc_ipu_hl_lib.c
+++ b/ipu/mxc_ipu_hl_lib.c
@@ -1236,12 +1236,6 @@ again:
 			}
 		}
 
-		if (ipu_priv_handle->output.fb_chan == MEM_FG_SYNC) {
-			if ( ioctl(ipu_priv_handle->output.fd_fb, MXCFB_SET_OVERLAY_POS,
-						&(output->fb_disp.pos)) < 0)
-				dbg(DBG_ERR, "Set FB position failed!\n");
-		}
-
 		dbg(DBG_INFO, "fb xres %d\n", fb_var.xres);
 		dbg(DBG_INFO, "fb yres %d\n", fb_var.yres);
 		dbg(DBG_INFO, "fb xres_virtual %d\n", fb_var.xres_virtual);
@@ -1307,6 +1301,13 @@ again:
 		if ( ioctl(ipu_priv_handle->output.fd_fb, FBIOBLANK, blank) < 0) {
 			dbg(DBG_ERR, "UNBLANK FB failed!\n");
 		}
+
+		if (ipu_priv_handle->output.fb_chan == MEM_FG_SYNC) {
+			if ( ioctl(ipu_priv_handle->output.fd_fb, MXCFB_SET_OVERLAY_POS,
+						&(output->fb_disp.pos)) < 0)
+				dbg(DBG_ERR, "Set FB position failed!\n");
+		}
+
 	}
 err:
 	return ret;
@@ -2627,7 +2628,6 @@ static void _mxc_ipu_lib_task_uninit(ipu_lib_priv_handle_t * ipu_priv_handle, pi
 
 	if (ipu_priv_handle->output.show_to_fb) {
 		if (ipu_priv_handle->output.fb_chan == MEM_FG_SYNC) {
-			struct mxcfb_pos pos = {0, 0};
 			struct fb_fix_screeninfo fb_fix;
 			struct fb_var_screeninfo fb_var;
 
@@ -2638,10 +2638,6 @@ static void _mxc_ipu_lib_task_uninit(ipu_lib_priv_handle_t * ipu_priv_handle, pi
 				__fill_fb_black(ipu_priv_handle->output.ofmt, &fb_var, &fb_fix,
 						ipu_priv_handle->output.fb_mem);
 			}
-
-			if ( ioctl(ipu_priv_handle->output.fd_fb, MXCFB_SET_OVERLAY_POS,
-						&pos) < 0)
-				dbg(DBG_ERR, "Set FB position failed!\n");
 		}
 	}
 
-- 
1.8.0

